Add separate non-TensorFlow test run in Travis CI

orionr opened the pull request:
Continuation of work for https://github.com/tensorflow/tensorboard/issues/2027 and ultimately removing the `TENSORBOARD_NO_TF=1` environment variable.

Plugins tests aren't run yet, but this confirms that build and limited tests run without TensorFlow installed.

wchargin commented on tensorboard/pip_package/build_pip_package.sh:
```diff
@@ -125,15 +125,22 @@ smoke() {
   pip install -qU pip
-  pip install -qU "${smoke_tf}"
+
+  if [ ! -z "${smoke_tf}" ]; then
```
Can just be `[ -n "${smoke_tf}" ]`.

wchargin commented on tensorboard/pip_package/build_pip_package.sh:
```diff
@@ -159,26 +166,32 @@ from tensorboard.plugins.beholder import Beholder, BeholderHook
   }
-  if is_tf_2 ; then
-    test_tf_summary ''
+  if is_tf ; then
```
Is there a reason that this isn’t just `if [ -n "${smoke_tf}" ]`? We
need the active check for TF 2.0 because we can’t easily discern the
TensorFlow major version from the Pip distribution specification, but
just checking whether we installed TensorFlow at all seems easy enough.

orionr replied:
Good point - changing.

wchargin commented on tensorboard/main.py:
```diff
@@ -51,12 +51,6 @@ def run_main():
-    # TENSORBOARD_NO_TF environment variable, we check for TensorFlow here so that if it's
-    # missing we generate a clear and immediate error rather than partial functionality.
-    # TODO(#2027): Remove environment check once we have placeholder UI
-    if os.getenv('TENSORBOARD_NO_TF') is None:
```
If this removal is included in this PR, then we can’t merge this PR
until #2032 is merged. If you don’t want to block on that PR, then it
looks to me like you should be able to add `TENSORBOARD_NO_TF=1` to the
Travis CI script and/or Pip package smoke test as appropriate. Or, feel
free to keep this as is if you don’t mind waiting for #2032.

orionr replied:
For now I'll keep it, but we can remove it if this is ready to land before the UI one. Thanks!

orionr commented:
The script works for me, but I'm not sure if I'm running it correctly. From a conda environment:

```
(notf) orionr-mbp:tensorboard orionr$ bazel run //tensorboard/pip_package:build_pip_package -- --tf-version ''
INFO: Analysed target //tensorboard/pip_package:build_pip_package (2 packages loaded, 760 targets configured).
INFO: Found 1 target...
Target //tensorboard/pip_package:build_pip_package up-to-date:
  bazel-bin/tensorboard/pip_package/build_pip_package
INFO: Elapsed time: 0.747s, Critical Path: 0.10s
INFO: 0 processes.
INFO: Build completed successfully, 4 total actions
INFO: Build completed successfully, 4 total actions
+ command -v curl
+ command -v perl
+ command -v virtualenv
```

wchargin commented:
It looks like it’s failing on Travis with the same error:

```
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/tmp/tensorboard/smoke-venv/local/lib/python2.7/site-packages/tensorboard/plugins/scalar/summary.py", line 89, in pb
    import tensorflow.compat.v1 as tf
ImportError: No module named tensorflow.compat.v1
The command "bazel run //tensorboard/pip_package:build_pip_package -- --tf-version "${TF_VERSION_ID}" --smoke" exited with 1.
```

Build: <https://travis-ci.com/tensorflow/tensorboard/jobs/189495702>

Your Bazel invocation is correct. Maybe something’s the matter in your
environment (either Conda or POSIX). I can reproduce the problem with

```
$ tmpdir="$(mktemp -d)"
$ virtualenv -q -p python2 "${tmpdir}/ve"
$ . "${tmpdir}/ve/bin/activate"
(ve) $ pip install -q "absl-py>=0.7.0" "numpy<2.0,>=1.14.5" boto3==1.9.86 flake8==3.5.0 futures==3.1.1 grpcio==1.6.3 mock==2.0.0 moto==1.3.7 yamllint==1.5.0
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
(ve) $ git status 
HEAD detached at orionr/run-notf-travis-tests
nothing to commit, working tree clean
(ve) $ git rev-parse --verify HEAD
3046da351a13e90b979227d564b287e878e9e41d
(ve) $ bazel run //tensorboard/pip_package:build_pip_package -- --tf-version ''
<snip; same traceback as in Travis>
```

in case that helps. Can you look into this when you get a chance?

I’m not sure what’s up with the `assertItemsEqual` Python 3 failures in
your Travis build, though. I can’t reproduce them, and they don’t make
much sense to me.

wchargin commented on .travis.yml:
```diff
@@ -127,7 +136,13 @@ script:
   - bazel build //tensorboard/...
-  - bazel test //tensorboard/...
+  - |
+    # Limit tests without TensorFlow
```
Small change requested: this shows up in failed builds as

<pre><code><strong>The command "# Limit tests without TensorFlow</strong>
if [ -z "${TF_VERSION_ID}" ]; then
  bazel test //tensorboard/... --test_tag_filters=support_notf
else
  bazel test //tensorboard/...
fi
" exited with 3.</code></pre>

which is a bit less readable than it used to be:

<pre><code><strong>The command "bazel test //tensorboard/..." exited with 3.</strong></code></pre>

Could you please replace this with something like

```yaml
  - |
    # When TensorFlow is not installed, run a restricted subset of tests.
    if [ -z "${TF_VERSION_ID}" ]; then
      test_tag_filters=support_notf
    else
      test_tag_filters=
    fi
  - bazel test //tensorboard/... --test_tag_filters="${test_tag_filters}"
```

so that the actual step that fails just has a `bazel test`?

orionr replied:
Good idea - changing.

wchargin commented on tensorboard/manager_test.py:
```diff
@@ -23,10 +23,11 @@
 
 import six
-import tensorflow as tf
+import unittest
```
Please revert changes to this file, as described above.

orionr replied:
Reverting test changes - I'll put those to the side for another PR.

wchargin commented on tensorboard/backend/application_test.py:
```diff
@@ -29,7 +29,7 @@
 
 import six
-import tensorflow as tf
+import unittest
```
Please revert changes to this file, as described above.

orionr commented:
Rebased and removed test changes, etc. However, I still can't repro the `ImportError: No module named tensorflow.compat.v1` error locally, which is weird. Going to look at that tomorrow. Thanks @wchargin for the review.

orionr commented:
@wchargin CI passes now! Let me know your thoughts and thanks.

wchargin commented on tensorboard/pip_package/build_pip_package.sh:
```diff
@@ -153,32 +161,41 @@ smoke() {
-    test_tf_summary ''
+  if [ -n "${smoke_tf}" ]; then
+    # Exhaustively test various sequences of importing tf.summary
+    # but only with TF
```
The change to this comment seems redundant to me. It’s clear both that
we _are_ only testing with TF (the immediately preceding line is the
guard) and that we _should_ only be testing with TF (it’s `tf.summary`
imports, after all). Suggest reverting.

wchargin commented:
Thanks!
